### ISSUE TITLE
Error when attack_idx is a 0-dim tensor

### ISSUE THREAD

**Issue by JiaDingCN** [bhheo/BSS_distillation#2]
Hi, thank you for your work. I am reproducing your code and find out it will report error when attach_idx is a 0-dim tensor. I simply modify the code from:

```python
if attack_idx.shape[0] > attack_size:
                    diff = (F.softmax(out_t[attack_idx,:], 1).data - F.softmax(out_s[attack_idx,:], 1).data) ** 2
                    distill_score = diff.sum(dim=1) - diff.gather(1, targets[attack_idx].data.unsqueeze(1)).squeeze()
                    attack_idx = attack_idx[distill_score.sort(descending=True)[1][:attack_size]]
```

to

```python
if attack_idx.dim()>0 and attack_idx.shape[0] > attack_size:
                    diff = (F.softmax(out_t[attack_idx,:], 1).data - F.softmax(out_s[attack_idx,:], 1).data) ** 2
                    distill_score = diff.sum(dim=1) - diff.gather(1, targets[attack_idx].data.unsqueeze(1)).squeeze()
                    attack_idx = attack_idx[distill_score.sort(descending=True)[1][:attack_size]]
```
 and from:

```python
class_score, class_idx = F.softmax(out_t, 1)[attack_idx, :].data.sort(dim=1, descending=True)
class_score = class_score[:, 1:]
class_idx = class_idx[:, 1:]

rand_seed = 1 * (class_score.sum(dim=1) * torch.rand([attack_idx.shape[0]]).cuda()).unsqueeze(1)
prob = class_score.cumsum(dim=1)
for k in range(attack_idx.shape[0]):
    for c in range(prob.shape[1]):
        if (prob[k, c] >= rand_seed[k]).cpu().numpy():
             attack_class[k] = class_idx[k, c]
             break
```

to:

```python
               if attack_idx.dim()==0:
                    
                    class_score, class_idx = F.softmax(out_t, 1)[attack_idx, :].data.sort(dim=0, descending=True)
                    class_score = class_score[1:]
                    class_idx = class_idx[1:]
                    rand_seed = 1 * (class_score.sum(dim=0) * torch.rand([1]).cuda()).unsqueeze(1)
                    prob = class_score.cumsum(dim=0)
                    k=0
                    for c in range(prob.shape[0]):
                        if (prob[c] >= rand_seed[k]).cpu().numpy():
                            attack_class = class_idx[c]
                            break
                else:
                    class_score, class_idx = F.softmax(out_t, 1)[attack_idx, :].data.sort(dim=1, descending=True)
                    class_score = class_score[:, 1:]
                    class_idx = class_idx[:, 1:]
                    rand_seed = 1 * (class_score.sum(dim=1) * torch.rand([attack_idx.shape[0]]).cuda()).unsqueeze(1)
                    prob = class_score.cumsum(dim=1)
                    for k in range(attack_idx.shape[0]):
                        for c in range(prob.shape[1]):
                            if (prob[k, c] >= rand_seed[k]).cpu().numpy():
                                attack_class[k] = class_idx[k, c]
                                break
```

Is that correct?

**Comment by bhheo** [comment]
The attack codes only work when attack_flag has a non-zero element

https://github.com/bhheo/BSS_distillation/blob/586b1411ed067af92ab781ac4fc58f8cf58798dc/train_BSS_distillation.py#L111

I can't understand the situation where the sum of `attack_flag` is larger than 0, but there is no dimension of `attack_idx`.
Could you give me an example with `attack_flag` and `attack_idx`?

**Comment by JiaDingCN** [comment]
Yes, when there is only 'True' in attack_flag, `attack_flag.sum()` will still be true, but `attack_idx = attack_flag.nonzero().squeeze()` will be a LongTensor, which does not have shape and the dimension is 0

(This situation occurs when I apply your method to cifar100)

Like this:

attack_flag:
```
tensor([False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False,  True, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False], device='cuda:0')
```
attack_idx:
```
tensor(197, device='cuda:0')
```
attack_idx.dim():
```
0
```
attack_idx.shape:
```
torch.Size([])
```

**Comment by bhheo** [comment]
Oh. I understand

I recommend to replace `squeeze()` function to `flatten()` function.

So, in here,
https://github.com/bhheo/BSS_distillation/blob/586b1411ed067af92ab781ac4fc58f8cf58798dc/train_BSS_distillation.py#L113
change squeeze to flatten
```python
attack_idx = attack_flag.nonzero().flatten() 
```
I think it might solve the problem. No zero dimension vectors from `flatten()` function.

**Comment by bhheo** [comment]
I think your revision also can be the solution to this problem.
But, I lost the setting for this code. I'm hard to verify it.

**Comment by JiaDingCN** [comment]
I just tried and changing squeeze to flatten works. Thank you.
